aws-lc-rs/src/buffer.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+//! This module exposes a buffer type used in crate APIs returning private keys and other "private"
+//! contents.
+
+#![allow(clippy::module_name_repetitions)]
+
+use std::borrow::Cow;
+use std::fmt;
+use std::marker::PhantomData;
+
+use zeroize::Zeroize;
+
+/// This is a buffer type for some data exposed by various APIs in this crate.
+///
+/// `T` acts as a discriminant between different kinds of data.
+///
+/// The buffer will be zeroed on drop if it is owned.
+pub struct Buffer<'a, T>(Cow<'a, [u8]>, PhantomData<T>);
+
+impl<'a, T> Drop for Buffer<'a, T> {
+    fn drop(&mut self) {
+        if let Cow::Owned(b) = &mut self.0 {
+            b.zeroize();
+        }
+    }
+}
+
+impl<'a, T> Buffer<'a, T> {
+    pub(crate) fn new(owned: Vec<u8>) -> Buffer<'a, T> {
+        Buffer(Cow::Owned(owned), PhantomData)
+    }
+
+    pub(crate) fn take_from_slice(slice: &mut [u8]) -> Buffer<'a, T> {
+        let owned = slice.to_vec();
+        slice.zeroize();
+        Buffer(Cow::Owned(owned), PhantomData)
+    }
+
+    pub(crate) fn public_from_slice(slice: &[u8]) -> Buffer<'_, T> {
```

Comment:
nit: Maybe `borrow_from_slice`. The take function is clear cause its going to copy and clear. This is more borrowing where the lifetime is bound to that of the passed in slice. But not sure on the choice of `public`.

Reply:
Public is intended to say that callers shouldn't use this for private material, mostly to push towards making sure take_from_slice is used where needed (otherwise easy to get this wrong).

Reply:
This isn't obvious to the reader.  Adding method doc should help.